### DISCUSSION TITLE
How can I get the PID of a background process in a Xonsh script?

### DISCUSSION THREAD

**Discussion by ading2210** [xonsh/xonsh#5310] [Q&A] [answered]
This is trivial in bash with the [`$!` variable](https://stackoverflow.com/questions/1908610/how-to-get-process-id-of-background), but I couldn't figure out how to do this in Xonsh.

Something like `!(sleep 1 &)` doesn't work, as waits until the command completes before returning anything. 

If this isn't possible, then how should I start a background process then kill it later without resorting to Python's `subprocess` library?

**Comment by ading2210** [comment, marked as answer]
I've found a workaround for this using the `xonsh.jobs` module:

```xsh
import xonsh

def get_last_job():
  jobs = list(xonsh.jobs.get_jobs().values())
  if not jobs:
    return None
  last_job = max(jobs, key=lambda x: x["started"])
  return last_job

sleep 10 &
j = get_last_job()

kill @(j["pids"][0])
kill @(j["pids"][0]) #no such process
```

**Reply by ading2210** [reply]
Note that in the latest Xonsh versions, `xonsh.jobs` has moved to `xonsh.procs.jobs`. You can use this code to detect this at runtime:
```python
import xonsh
if hasattr(xonsh, "jobs"):
  xonsh_jobs = xonsh.jobs
else:
  xonsh_jobs = xonsh.procs.jobs
```

And then change get_last_job() to reference `xonsh_jobs` instead. 
```python
def get_last_job():
  jobs = list(xonsh_jobs.get_jobs().values())
  if not jobs:
    return None
  last_job = max(jobs, key=lambda x: x["started"])
  return last_job
```

**Reply by anki-code** [reply]
Feel free to add this to [xonsh.api](https://github.com/xonsh/xonsh/tree/main/xonsh/api) e.g. `from xonsh.api.jobs import get_last_job`. And then add the example to [bash to xonsh guide](https://xon.sh/bash_to_xsh.html).
